autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,351 @@
+            )]
+        )
+
+    def get_last_epoch(self) -> int:
+        if 'train_loss' not in self.performance_tracker:
+            return 0
+        else:
+            return max(self.performance_tracker['train_loss'].keys())
+
+    def repr_last_epoch(self) -> str:
+        """
+        For debug purposes, returns a nice representation of last epoch
+        performance
+
+        Returns:
+            str: A nice representation of the last epoch
+        """
+        last_epoch = len(self.performance_tracker['train_loss'])
+        string = "\n"
+        string += '=' * 40
+        string += f"\n\t\tEpoch {last_epoch}\n"
+        string += '=' * 40
+        string += "\n"
+        for key, value in sorted(self.performance_tracker.items()):
+            if isinstance(value[last_epoch], dict):
+                # Several metrics can be passed
+                string += "\t{}:\n".format(
+                    key,
+                )
+                for sub_key, sub_value in sorted(value[last_epoch].items()):
+                    string += "\t\t{}: {}\n".format(
+                        sub_key,
+                        sub_value,
+                    )
+            else:
+                string += "\t{}: {}\n".format(
+                    key,
+                    value[last_epoch],
+                )
+        string += '=' * 40
+        return string
+
+
+class BaseTrainerComponent(autoPyTorchTrainingComponent):
+
+    def prepare(
+        self,
+        metrics: List[Any],
+        model: torch.nn.Module,
+        criterion: torch.nn.Module,
+        budget_tracker: BudgetTracker,
+        optimizer: torch.nn.Module,
+        device: torch.device,
+        logger: logging.Logger,
+        writer: Optional[SummaryWriter],
+    ) -> None:
+
+        self.logger = logger
+
+        # Save the device to be used
+        self.device = device
+
+        # Setup the metrics
+        self.metrics = metrics
+
+        # Setup the loss function
+        self.criterion = criterion.to(device)
+
+        # setup the model
+        self.model = model.to(device)
+
+        # setup the optimizers
+        self.optimizer = optimizer
+
+        # The budget tracker
+        self.budget_tracker = budget_tracker
+
+        # A summary writer for tensorboard
+        self.writer = writer
+
+    def on_epoch_start(self, X: Dict[str, Any], epoch: int) -> None:
+        """
+        Optional place holder for AutoPytorch Extensions.
+
+        An user can define what happens on every epoch start or every epoch end.
+        """
+        pass
+
+    def on_epoch_end(self, X: Dict[str, Any], epoch: int) -> bool:
+        """
+        Optional place holder for AutoPytorch Extensions.
+        An user can define what happens on every epoch start or every epoch end.
+        If returns True, the training is stopped
+
+        """
+        return False
+
+    def train(self, train_loader: torch.utils.data.DataLoader, epoch: int
```

Comment:
Maybe a better name would be train_epoch or something indicating that this is one epoch

Reply:
thanks for the suggestion I have renamed this procedure!